### PR TITLE
Unmatched n singletons reads handling

### PR DESCRIPTION
**Pull request by vinjana**
#2 : Unmatched reads were not handled correctly. This should implement this feature.

### CODE REVIEW

**Comment by vinjana** [comment]
Tested on an in-house dataset with unmatched reads.

File: resources/analysisTools/bam2fastq/coreutilsSortFastqPair.sh [hunk 1]
```diff
@@ -139,6 +139,8 @@ fi
 
 waitForAll_BashSucksVersion
 
+sleep 5   # Wait for network filesystem delays
```
Comment (dankwart-de) [inline resources/analysisTools/bam2fastq/coreutilsSortFastqPair.sh]: Any better idea? Sleep can be very hazardous... though I do not know really know of a better solution.

Comment (dankwart-de) [reply resources/analysisTools/bam2fastq/coreutilsSortFastqPair.sh]: You could try to wait for the files instead.

File: src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy [hunk 2]
```diff
@@ -172,20 +162,23 @@ class BamToFastqWorkflow extends Workflow {
...
+                    for (ReadGroup grp : groupsWithUnsortedFastqs.readGroups) {
+                        sortPairedFastqs(bamFile.getPath().name, grp.name, grp[ReadFileType.READ1], grp[ReadFileType.READ2]) // R1/R2
+                        sortFastq(bamFile.getPath().name, grp.name, grp[ReadFileType.UNMATCHED_READ1]) // U1
+                        sortFastq(bamFile.getPath().name, grp.name, grp[ReadFileType.UNMATCHED_READ2]) // U2
+                        sortFastq(bamFile.getPath().name, grp.name, grp[ReadFileType.SINGLETON]) // S
                     }
                 }
 
             } else {
 //                FileGroup out = bam2fastq(cfg, BaseFile)
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy:197]: Remove comments

Comment (dankwart-de) [reply src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy:197]: Maybe just state, that this alternative is not implemented yet? You could check that earlier and skip execution at all.

File: src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroupGroup.groovy [hunk 3]
```diff
@@ -0,0 +1,47 @@
+package de.dkfz.roddy.bam2fastq
+
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroupGroup.groovy:2]: Any tests for the class?

File: src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroup.groovy [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    private ReadFileType(String shortCut, Integer index) {
+        this.shortCut = shortCut
+        this.index = index
+    }
+
+}
+
+
+@CompileStatic
+class ReadGroup {
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroup.groovy]: Any tests for the class?

File: src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy [hunk 5]
```diff
@@ -205,7 +198,7 @@ class BamToFastqWorkflow extends Workflow {
     }
 
     protected boolean checkReadGroups(List<BaseFile> bamFiles) {
-        Map<String, List<String>> readGroups = bamFiles.collectEntries { [(it): readGroupsPerBamfile[it.absolutePath]] }
+        Map<String, ReadGroupGroup> readGroups = bamFiles.collectEntries { [(it): readGroupsPerBamfile[it.absolutePath]] }
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy]: Any tests for the class?

File: resources/analysisTools/bam2fastq/coreutilsSortFastqPair.sh (hunk 1 above)
Comment (vinjana) [reply resources/analysisTools/bam2fastq/coreutilsSortFastqPair.sh]: Sleep can be hazardous?

Comment (vinjana) [reply resources/analysisTools/bam2fastq/coreutilsSortFastqPair.sh]: Yes, waiting for files may be more reasonable here. The only problem -- when being strict -- is thus I would introduce long-range dependencies in the code. For the process-IDs this is solved by the registry that is checked in `waitForAll_BashSucksVersion` function.

Comment (dankwart-de) [reply resources/analysisTools/bam2fastq/coreutilsSortFastqPair.sh]: Ok, then leave the sleep. Infrastructure problems suck.

Comment (dankwart-de) [reply resources/analysisTools/bam2fastq/coreutilsSortFastqPair.sh]: Sleep itself is not hazardous. But it also won't guarantee, that the files exist then.

File: src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy [hunk 6]
```diff
@@ -162,12 +179,12 @@ class BamToFastqWorkflow extends Workflow {
         for (BaseFile bamFile : bamFiles) {
 
             if (config.outputPerReadGroup) {
-                ReadGroupGroup groups = readGroupsPerBamfile[bamFile.absolutePath]
+                ReadGroupGroup<BaseFile> groups = readGroupsPerBamfile[bamFile.absolutePath]
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy:183]: If you define ReadGroupGroup with BaseFile like <? extends BaseFile>, you can omit it here.

File: src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroup.groovy [hunk 7]
```diff
@@ -23,19 +23,19 @@ enum ReadFileType {
 
 
 @CompileStatic
-class ReadGroup {
+class ReadGroup<T> {
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroup.groovy:26]: As stated above, you could use T extends BaseFile or so.

File: src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroupGroup.groovy [hunk 8]
```diff
@@ -1,33 +1,18 @@
 package de.dkfz.roddy.bam2fastq
 
-import de.dkfz.roddy.knowledge.files.FileGroup
+
 import groovy.transform.CompileStatic
 
+/** Streamlined handling of a set of ReadGroups. */
 @CompileStatic
-class ReadGroupGroup {
+class ReadGroupGroup<T> {
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroupGroup.groovy:8]: As said, T extends BaseFile?

File: src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy [hunk 9]
```diff
@@ -82,13 +82,30 @@ class BamToFastqWorkflow extends Workflow {
         readGroupsPerBamfile[bamfileName]
     }
 
+    /** Update the ReadGroups with the names of the files as provided in a FileGroup. */
+    private ReadGroupGroup<BaseFile> updatedReadGroupGroup(final ReadGroupGroup<BaseFile> oldGroups, final FileGroup fileGroup) {
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy:88]: Why did you move this away from the readgroup? I don't think it should be located in the workflow. Then maybe use a helper class or so.

File: src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroup.groovy [hunk 10]
```diff
@@ -47,15 +47,15 @@ class ReadGroup {
     }
 
     /** Functional interface function doing a (shallow-)copy-on-write */
-    ReadGroup updatedFile(ReadFileType type, BaseFile newFile) {
-        Map<ReadFileType, BaseFile> newFiles = files.collectEntries { iterType, oldFile ->
+    ReadGroup<T> updatedFile(ReadFileType type, T newFile) {
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroup.groovy:50]: Here you still have the update method in place.

File: src/test/de/dkfz/roddy/bam2fastq/ReadGroupGroupTest.groovy [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        def rgB = new ReadGroup<String>("b")
+        def rgg1 = new ReadGroupGroup([])
+        def rgg2 = new ReadGroupGroup([rgA, rgB])
+
+        then:
+        rgg1.readGroups == []
+        rgg2.readGroups == [rgA, rgB]
+    }
+
+    def "SetReadGroups"() {
```
Comment (dankwart-de) [inline src/test/de/dkfz/roddy/bam2fastq/ReadGroupGroupTest.groovy]: Put exception in separate method or rename this?

File: src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy (hunk 9 above)
Comment (vinjana) [reply src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy:88]: It does not belong into ReadGroupGroup either, as this is generic with unconstrained T but this code is specific for T extends BaseFile. I was thinking of a subclass, but that did not seem to be much better. The problem is also the FileGroup class, that has nothing to do with the ReadGroups. In this sense, this code does actually not belong to ReadGroupGroup.

Comment (vinjana) [reply src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy:88]: Marked as deprecated.

File: src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroup.groovy (hunk 10 above)
Comment (vinjana) [reply src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroup.groovy:50]: This is generic it unconstrained T and thus a simple copys-on-write update method/copy construction.

File: src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroupGroup.groovy (hunk 8 above)
Comment (vinjana) [reply src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroupGroup.groovy:8]: Nope, otherwise the tests get more complex.

File: src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroup.groovy (hunk 7 above)
Comment (vinjana) [reply src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroup.groovy:26]: Nope, otherwise the tests get more complex.

File: src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy (hunk 6 above)
Comment (vinjana) [reply src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy:183]: see below

File: src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy (hunk 5 above)
Comment (vinjana) [reply src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy]: Done.

File: src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroup.groovy (hunk 4 above)
Comment (vinjana) [reply src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroup.groovy]: Done.

File: src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroupGroup.groovy (hunk 3 above)
Comment (vinjana) [reply src/main/groovy/de/dkfz/roddy/bam2fastq/ReadGroupGroup.groovy:2]: done

**Review by dankwart-de — APPROVED** [review]
